src/Metric.jl

```diff
@@ -1,50 +1,202 @@
+"""
+function christofell_symbols_first(M::MetricManifold, x)
     n = size(x, 1)
-    ginv = inverse_local_metric(M, x)
     ∂g = Array(reshape(ForwardDiff.jacobian(x -> local_metric(M, x), x), n, n, n))
-    @einsum Γ[k,i,j] := 1/2 * ginv[k,l] * (∂g[l,j,i] + ∂g[l,i,j] - ∂g[i,j,l])  # Absil et al, Eq 5.12
+    @tensor Γ[i,j,k] := 1/2 * (∂g[k,j,i] + ∂g[i,k,j] - ∂g[i,j,k])
     return Γ
 end
 
-function exp!(M::MetricManifold, y, x, v)
+@doc doc"""
+    christofell_symbols_second(M::MetricManifold, x)
+
+Compute the Christofell symbols of the second kind in local coordinates.
+The Christofell symbols are (in Einstein summation convention)
+
+$\Gamma^{\ell}_{ij} = \Gamma_{ijk} g^{k\ell},$
+
+where $\Gamma_{ijk}$ are the Christofell symbols of the first kind, and
+$g^{k\ell}$ is the inverse of the local representation of the metric tensor.
+"""
+function christofell_symbols_second(M::MetricManifold, x)
+    n = size(x, 1)
+    ginv = inverse_local_metric(M, x)
+    Γ₁ = christofell_symbols_first(M, x)
+    @tensor Γ₂[i,j,l] := ginv[k,l] * Γ₁[i,j,k]
+    return Γ₂
+end
+
+"""
+    riemann_tensor(M::MetricManifold, x)
+
+Compute the Riemann tensor, also known as the Riemann curvature tensor,
+at the point `x`.
+"""
+function riemann_tensor(M::MetricManifold, x)
+    n = size(x, 1)
+    Γ = christofell_symbols_second(M, x)
+    ∂Γ = Array(reshape(ForwardDiff.jacobian(x -> christofell_symbols_second(M, x), x), n, n, n, n))
+    @tensor R[i,j,k,l] := ∂Γ[i,k,l,j] - ∂Γ[i,j,l,k] + Γ[j,s,l] * Γ[i,k,s] - Γ[k,s,l] * Γ[i,j,s]
+    return R
+end
+
+"""
+    ricci_tensor(M::MetricManifold, x)
+
+Compute the Ricci tensor, also known as the Ricci curvature tensor,
+of the manifold `M` at the point `x`.
+"""
+function ricci_tensor(M::MetricManifold, x)
+    R = riemann_tensor(M, x)
+    @tensor Ric[i,j] := R[i,l,j,l]
+    return Ric
+end
+
+"""
+    ricci_curvature(M::MetricManifold, x)
+
+Compute the Ricci scalar curvature of the manifold `M` at the point `x`.
+"""
+function ricci_curvature(M::MetricManifold, x)
+    ginv = inverse_local_metric(M, x)
+    Ric = ricci_tensor(M, x)
+    S = sum(ginv .* Ric)
+    return S
+end
+
+"""
+    gaussian_curvature(M::MetricManifold, x)
+
+Compute the Gaussian curvature of the manifold `M` at the point `x`.
+"""
+gaussian_curvature(M::MetricManifold, x) = ricci_curvature(M, x) / 2
+
+"""
+    einstein_tensor(M::MetricManifold, x)
+
+Compute the Einstein tensor of the manifold `M` at the point `x`.
+"""
+function einstein_tensor(M::MetricManifold, x)
+    Ric = ricci_tensor(M, x)
+    g = local_metric(M, x)
+    ginv = inverse_local_metric(M, x)
+    S = sum(ginv .* Ric)
+    G = Ric - g .* S / 2
+    return G
+end
+
+function exp_diffeq_system!(du, u, p, t)
+    n = Int(size(u, 1) / 2)
+    M = p
+    x = view(u, 1:n)
+    dx = view(u, n+1:2n)
+    du[1:n] .= dx
+    Γ = christofell_symbols_second(M, x)
+    @tensor d²x[k] := -Γ[i,j,k] * dx[i] * dx[j]
+    du[n+1:end] .= d²x
+end
```

Comment:
In low-dimensional systems (dim<100), you may want to consider using the out-of-place form with signature `(u, p, t)`, returning `StaticVectors`, even though this may add another dependency. Not sure if performance is important here (like solving for many geodesics)?

Reply:
I think we should care, at least to some degree, about performance. `StaticArrays` will likely be added as a dependency anyway as just using `SizedArray` can provide big benefits for operations small arrays.

Reply:
I agree that we should support `SizedArray`s throughout the package. https://github.com/JuliaNLSolvers/ManifoldMuseum/pull/12/commits/6c3514bdb68a223bfc1cb49f99f91dca7e8f9607 and https://github.com/JuliaNLSolvers/ManifoldMuseum/pull/12/commits/5926eca563673be7d7596578a1cb3192d4438574 switches to the out-of-place form and supports sized arrays throughout, regardless of dimension. I'd appreciate any feedback.

We also should have a version of `exp` that takes a time vector and returns a vector of points corresponding to the curve. For numerically integrated geodesics, we're already computing the entire curve anyways, and this will be nice to have even for explicit geodesics.

Reply:
You mean a function like `geodesic(M,x,y)` that basically returns the function `t -> exp(M,x, t*log(M,x,y))`? one could then also define `geodesic(M,x,y,t)` to evaluate the geodesic at t and the short hand `geodesic(M,x,y,T)` for a vector `T` of time points that simply broadcasts on t.

Reply:
Something like that, but I was thinking for `exp` too, since we may not always have a `log` map. On that note, `geodesic` maybe should be renamed to `minimal_geodesic`, since that's more descriptive of what it means (both return geodesics, but `geodesic` returns specifically the distance-minimizing geodesic between the provided points).

Reply:
As long as we have a way to distinguish points and vectors, we could use `geodesic(M,x,y)` and `geodesic(M,x,v)`, otherwise we need two names. And yes, I am fine with adding the word `minimal`, though one could also use `shortest`.

Reply:
How would `geodesic(M,x,v)` in this conception differ from `exp(M,x,v)`?

Reply:
`exp(M,x,v)`returns a point `y`. `geodesic(M,x,v)` returns a function `t -> exp(M,x,t*v)`, so `geodesic(M,x,v)(1)` is the same as `exp` but one could again also implement `geodesic(M,x,v,T)` for a time vector `T` to get a point vector back.

Reply:
Got it. Hmm, but in many cases, we may not have a way to distinguish points from vectors, so what about this?
`shortest_geodesic(M,x,y)` returns `geodesic(M,x,log(M,x,y))`
`shortest_geodesic(M,x,y,t)` returns `geodesic(M,x,log(M,x,y),t)`
`shortest_geodesic(M,x,y,T)` returns `geodesic(M,x,log(M,x,y),T)`
`geodesic(M,x,v)` returns `t->exp(M,x,v,t)`
`geodesic(M,x,v,t)` returns `exp(M,x,v,t)`
`geodesic(M,x,v,T)` returns `exp(M,x,v,T)`
`exp(M,x,v)` computes exponential map
`exp(M,x,v,t)` returns `exp(M,x,t*v)`
`exp(M,x,v,T)` returns `[exp(M,x, t*v) for t in T]`

This makes behavior/arguments more transparent and doesn't require any additional overloads than the usual `exp!` functions. In some cases, a more efficient form of `exp(M,x,v,T)` might be possible, such as with the numerically integrated geodesics, and the user is free to implement that.

Reply:
At least in my head, any `geodesic(M,x,v)` would still be shortest (for example with length pi on the sphere you reach – not uniquely – the opposite point but all these are shortest geodesics). Thats why until now I called both (xy and xv geodesics) the same.

Having both is fine - I just think your default geodesics might be shortest, too.

Reply:
Geodesics are only locally distance-minimizing, so they don't guarantee the shortest distance between any two points on the geodesic. A shortest geodesic would be a (not necessarily unique) geodesic that does guarantee the shortest distance and would require a `log` map.

I'll give an example. Suppose we have points `x=(0,0,1)`, `y=(-1,0,0)` with `v=(1,0,0)`. `geodesic(M,x,v)` will give a geodesic that includes `x` and `y`, but it will take `t=3pi/4` time units to get to `y` tracing out an arc of the same length, whereas  `shortest_geodesic(M,x,y)` is equivalent to `geodesic(M,x,(-pi/4,0,0))` and will get to `y` at `t=1` tracing out an arc length of `pi/4`. And indeed, `pi/4` is the distance between `x` and `y`.

Reply:
Yes I know that minimizing only holds locally :) but I wasn't aware of your idea of modelling here, so thanks for the explanation, with that idea the naming is consistent and very good.

Reply:
Everything I said above, but replace `pi` with `tau=2pi`.

Reply:
Yeah. I also got the idea on the smaller sphere ;)

Reply:
Sorry, I commented that before I saw your comment :).